Mathlib/Data/Complex/Basic.lean

```diff
@@ -757,11 +757,57 @@ protected theorem mul_inv_cancel {z : ℂ} (h : z ≠ 0) : z * z⁻¹ = 1 := by
     ofReal_one]
 #align complex.mul_inv_cancel Complex.mul_inv_cancel
 
-/-! ### Field instance and lemmas -/
+noncomputable instance : RatCast ℂ where
+  ratCast := Rat.castRec
+
+/-! ### Cast lemmas -/
+
+@[simp, norm_cast]
+theorem ofReal_nat_cast (n : ℕ) : ((n : ℝ) : ℂ) = n :=
+  map_natCast ofReal n
+#align complex.of_real_nat_cast Complex.ofReal_nat_cast
 
+@[simp, norm_cast]
+theorem nat_cast_re (n : ℕ) : (n : ℂ).re = n := by rw [← ofReal_nat_cast, ofReal_re]
+#align complex.nat_cast_re Complex.nat_cast_re
+
+@[simp, norm_cast]
+theorem nat_cast_im (n : ℕ) : (n : ℂ).im = 0 := by rw [← ofReal_nat_cast, ofReal_im]
+#align complex.nat_cast_im Complex.nat_cast_im
+
+@[simp, norm_cast]
+theorem ofReal_int_cast (n : ℤ) : ((n : ℝ) : ℂ) = n :=
+  map_intCast ofReal n
+#align complex.of_real_int_cast Complex.ofReal_int_cast
+
+@[simp, norm_cast]
+theorem int_cast_re (n : ℤ) : (n : ℂ).re = n := by rw [← ofReal_int_cast, ofReal_re]
+#align complex.int_cast_re Complex.int_cast_re
+
+@[simp, norm_cast]
+theorem int_cast_im (n : ℤ) : (n : ℂ).im = 0 := by rw [← ofReal_int_cast, ofReal_im]
+#align complex.int_cast_im Complex.int_cast_im
+
+@[simp, norm_cast]
+theorem rat_cast_im (q : ℚ) : (q : ℂ).im = 0 := by
+  show (Rat.castRec q : ℂ).im = 0
+  cases q
+  simp [Rat.castRec]
+#align complex.rat_cast_im Complex.rat_cast_im
+
+@[simp, norm_cast]
+theorem rat_cast_re (q : ℚ) : (q : ℂ).re = (q : ℝ) := by
+  show (Rat.castRec q : ℂ).re = _
+  cases q
+  simp [Rat.castRec, normSq, Rat.mk_eq_divInt, Rat.mkRat_eq_div, div_eq_mul_inv, *]
+#align complex.rat_cast_re Complex.rat_cast_re
+
+/-! ### Field instance and lemmas -/
 
 noncomputable instance : Field ℂ :=
-{ inv := Inv.inv
+{ qsmul := fun n z => ⟨n • z.re - 0 * z.im, n • z.im + 0 * z.re⟩
```

Comment:
Please add a note here why we have the weird definition?

Reply:
I added a comment, but I actually don't understand the comment to do with `RestrictScalars` on the instance in `Data.Complex.Module`, so I don't fully understand the reason behind this change.

Reply:
I should maybe improve that comment as well.

Reply:
I just realised I could also change the definition of `Field` to this, and change the `ratCast` def to go through the reals. This also makes the instance diagram commute, but for a more complex reason.

```lean
noncomputable instance : Field ℂ :=
{ ratCast := fun q => (q : ℝ)
  ratCast_mk := fun a b h1 h2 => by
    show ofReal' _ = _
    simp [Rat.cast_mk' a b h1 h2]; rfl
  inv := Inv.inv
  mul_inv_cancel := @Complex.mul_inv_cancel
  inv_zero := Complex.inv_zero }
```